chainntnfs/txconfnotifier.go

```diff
@@ -98,78 +101,116 @@ func NewTxConfNotifier(startHeight uint32, reorgSafetyLimit uint32) *TxConfNotif
-		// the transaction has already been confirmed.
-		select {
-		case <-tcn.quit:
-			return fmt.Errorf("TxConfNotifier is exiting")
-		case ntfn.Event.Updates <- 0:
-		}
+	// First, we'll determine whether we have any active notifications for
+	// this transaction.
+	ntfns, ok := tcn.confNotifications[txid]
+	if !ok {
+		return errors.New("no active notification requests found")
+	}
 
-		select {
-		case <-tcn.quit:
-			return fmt.Errorf("TxConfNotifier is exiting")
-		case ntfn.Event.Confirmed <- txConf:
-			ntfn.dispatched = true
-		}
-	} else {
-		// Otherwise, we'll record the transaction along with the height
-		// at which we should notify the client.
-		ntfn.details = txConf
-		ntfnSet, exists := tcn.ntfnsByConfirmHeight[confHeight]
-		if !exists {
-			ntfnSet = make(map[*ConfNtfn]struct{})
-			tcn.ntfnsByConfirmHeight[confHeight] = ntfnSet
+	// Now, for each notification request, we'll attempt to dispatch a
+	// notification to the caller if the transaction has met its required
+	// number of confirmations.
+	for _, ntfn := range ntfns {
+		// Skip the notification request if a notification has already
+		// been dispatched.
+		if ntfn.dispatched {
+			continue
 		}
-		ntfnSet[ntfn] = struct{}{}
 
-		// We'll also send an update to the client of how many
-		// confirmations are left for the transaction to be confirmed.
-		numConfsLeft := confHeight - tcn.currentHeight
-		select {
-		case ntfn.Event.Updates <- numConfsLeft:
-		case <-tcn.quit:
-			return errors.New("TxConfNotifier is exiting")
+		confHeight := details.BlockHeight + ntfn.NumConfirmations - 1
+		if confHeight <= tcn.currentHeight {
+			Log.Infof("Dispatching %v conf notification for %v",
+				ntfn.NumConfirmations, ntfn.TxID)
+
+			// We'll send a 0 value to the Updates channel,
+			// indicating that the transaction has already been
+			// confirmed.
+			select {
+			case ntfn.Event.Updates <- 0:
+			case <-tcn.quit:
+				return errors.New("TxConfNotifier is exiting")
+			}
+
+			select {
+			case ntfn.Event.Confirmed <- details:
+				ntfn.dispatched = true
+			case <-tcn.quit:
+				return errors.New("TxConfNotifier is exiting")
+			}
+		} else {
+			// Otherwise, we'll record the transaction along with
+			// the height at which we should notify the client.
+			ntfn.details = details
+			ntfnSet, exists := tcn.ntfnsByConfirmHeight[confHeight]
+			if !exists {
+				ntfnSet = make(map[*ConfNtfn]struct{})
+				tcn.ntfnsByConfirmHeight[confHeight] = ntfnSet
+			}
+			ntfnSet[ntfn] = struct{}{}
+
+			// We'll also send an update to the client of how many
+			// confirmations are left for the transaction to be
+			// confirmed. We'll only do this if the transaction has
+			// been found in the chain before our current height to
+			// avoid sending duplicate updates.
+			if details.BlockHeight <= tcn.currentHeight {
+				numConfsLeft := confHeight - tcn.currentHeight
+				select {
+				case ntfn.Event.Updates <- numConfsLeft:
```

Comment:
Seems like there might be an issue here if multiple callers register for confirmation of the same txid. Each time `UpdateConfDetails` is called as a result of this, all clients will be sent updates? This will lead to each client receiving the updates multiple times, and the channel might even get full.

From what I can gather, we shouldn't need to iterate the ntfn clients in this method, but pass in the one client that requested the historical dispatch, and only notify that one.

Reply:
Yeah, I spent quite some time debugging this. Ended up adding an `id` so that we can uniquely track each notification. This is something that will be needed anyway once we support canceling confirmation notifications.